mypy/build.py

```diff
@@ -861,59 +863,136 @@ def is_file(path: str) -> bool:
+            user_dir = site.getusersitepackages()
+            return site.getsitepackages() + [user_dir]
+        # If site doesn't have get(user)sitepackages, we are running in a
+        # virtualenv, and should fall back to get_python_lib
+        return [get_python_lib()]
+    else:
+        # Use subprocess to get the package directory of given Python
+        # executable
+        try:
+            output = call_python(python_executable, USER_SITE_PACKAGES)
+        except subprocess.CalledProcessError:
+            # if no paths are found (raising a CalledProcessError), we fall back on sysconfig,
+            # the python executable is likely in a virtual environment, thus lacking
+            # needed site methods
+            output = call_python(python_executable, VIRTUALENV_SITE_PACKAGES)
+        return [line for line in output.splitlines() if os.path.isdir(line)]
+
+
+def find_base_dirs(lib_path: Tuple[str, ...], dir_chain: str, components: List[str],
+                   site_packages_dirs: List[str]) -> Tuple[Iterable[str], List[str]]:
+    # If we're looking for a module like 'foo.bar.baz', it's likely that most of the
+    # many elements of lib_path don't even have a subdirectory 'foo/bar'.  Discover
+    # that only once and cache it for when we look for modules like 'foo.bar.blah'
+    # that will require the same subdirectory.
+    dirs = find_module_dir_cache.get((dir_chain, lib_path), [])
+    if not dirs:
+        # Regular packages on the PATH
+        for pathitem in lib_path:
+            # e.g., '/usr/lib/python3.4/foo/bar'
+            isdir = find_module_isdir_cache.get((pathitem, dir_chain))
+            if isdir is None:
+                dir = os.path.normpath(os.path.join(pathitem, dir_chain))
+                isdir = os.path.isdir(dir)
+                find_module_isdir_cache[pathitem, dir_chain] = isdir
+            if isdir:
+                dirs.append(dir)
+        find_module_dir_cache[dir_chain, lib_path] = dirs
+    third_party_dirs = []
+    # Third-party stub/typed packages
+    for pkg_dir in site_packages_dirs:
+        prefix = components[0]
+        stub_name = components[0] + '-stubs'
+        typed_file = os.path.join(pkg_dir, components[0], 'py.typed')
+        stub_dir = os.path.join(pkg_dir, stub_name)
+        if os.path.isdir(stub_dir):
+            components[0] = stub_name
+            rest = components[:-1]
+            path = os.path.join(pkg_dir, *rest)
+            if os.path.isdir(path):
+                third_party_dirs.append(path)
+            components[0] = prefix
+        elif os.path.isfile(typed_file):
+            path = os.path.join(pkg_dir, dir_chain)
+            third_party_dirs.append(path)
+
+    return tuple(third_party_dirs +
+                 find_module_dir_cache[dir_chain, lib_path]), components
+
+
+def find_module_in_base_dirs(id: str, candidate_base_dirs: Iterable[str],
+                             last_component: str) -> Optional[str]:
```

Comment:
This isn't really correct on python 2 - you need to walk up the tree looking for __init__.py files at each level, rather than assuming that only the last stage matters

Reply:
Note: this is one of the fundamental correctness issues for mypy's import implementation (that should be addressed in a separate PR, not here). It's not even enough to verify `__init__.py` at each level, you really have to go level-by-level from the top down, otherwise if you are importing `a.b.c` and you have two different `a/__init__.py` on the path, you might "find" `a.b.c` in the wrong one and think everything is fine, when really the code you found is not importable at runtime.